I87 six

jamiebull1 opened the pull request:
This implements the alternative suggestion for Python 3 support in #87. This also makes the CI run much quicker since it doesn't need to run `2to3` and means that we can have a single codebase for python 2 and python 3.

It also fixes #64 by using environment markers in the `requirements.txt` file to install the correct version of required packages. For example:

    pydot>1.0; python_version <= '2.7'
    pydot3k; python_version >= '3.0'

jamiebull1 commented on build/lib/eppy/Main_Tutorial.py:
```diff
@@ -129,7 +129,7 @@
 
```
This tutorial needs bringing into line with `six` so it can run in either 2 or 3.

jamiebull1 commented on build/lib/eppy/bunchhelpers.py:
```diff
@@ -48,10 +48,10 @@ def replaceint(fname, replacewith='%s'):
 def cleaniddfield(acomm):
     """make all the keys lower case"""
-    for key in list(acomm.keys()):
+    for key in acomm.keys():
```
This looks like it isn't being tested, since it should fail in Python 3.

jamiebull1 commented on build/lib/eppy/bunchhelpers.py:
```diff
@@ -48,10 +48,10 @@ def replaceint(fname, replacewith='%s'):
         val = acomm[key]
         acomm[key.lower()] = val
-    for key in list(acomm.keys()):
+    for key in acomm.keys():
```
Same here

jamiebull1 commented on build/lib/eppy/ex_loopdiagram.py:
```diff
@@ -71,7 +71,7 @@ def dropnodes(edges):
                     try:
                         newedges.index(newtup)
-                    except ValueError as e:
+                    except ValueError, e:
```
Also not valid in 3

jamiebull1 commented:
Ok, having started to go through the code, there's quite a bit that's still not version-agnostic in areas that don't have test coverage. I'll try and find time to go over them and bring them into line.

jamiebull1 commented:
Ok, time found. This should be ready to go now, @santoshphilip

jamiebull1 commented:
@santoshphilip I'll try to keep this up-to-date with the `develop` branch as you merge things in so it's ready when you need it. So far the only conflict I've had is needing to remove the /p3 folder.

santoshphilip commented:
@jamiebull1 Do you think the six branch is ready for prime time ?
Or should we wait until we have greater test coverage.

jamiebull1 commented:
@santoshphilip I think it's ready for `develop` at least. A lot of the reason test coverage doesn't look great at the moment is because we're including things that shouldn't be counted (unused/old versions of functions that should probably be cleaned up, etc.). I'd be happier if things that are ready to go into `develop` were merged in before focusing on getting the coverage up though.

There's also a fair bit of improvement to the CI on #101 including testing for Python 2 and 3 on OSX as well as on Windows and Linux, and some things to improve test coverage [(link to branch in my repo)](https://codecov.io/gh/santoshphilip/eppy/branch/i62_eplusrunner) that gets it up to around 80% and runs much faster. Might be worth taking a look at that when you get time, and seeing if between us we can figure out the best way to separate out unit tests from integration tests. It would leave a much smaller target to aim at for getting the rest of the code unit tested.

santoshphilip commented:
Great.I'll plan on six being in develop branch

jamiebull1 commented:
Looks like this is OK now. I think it's ready for use in `develop`, but could do with a bit more test coverage before it goes into `master`, particularly because quite a bit has changed in areas which have no coverage at all at the moment.

lafrech commented:
Hi.

I thought you'd appreciate a little feedback.

We just ran 

    pip install git+https://github.com/santoshphilip/eppy@i87_six

and everything went smooth.

Installation worked (I saw a few of [those deprecation warnings](https://github.com/pypa/pip/blob/80fedf73346d8a06572e726625cd3a12467c6932/pip/req/req_install.py#L606), but I can't tell whether or not they're related to eppy itself) and we didn't see any regression in our application.

Thank you and congratulations for this.

jamiebull1 commented:
@lafrech Thanks for testing this out. Glad to hear it worked for you. If at all possible, I'd really appreciate the same test on

     pip install git+https://github.com/jamiebull1/eppy@improve_coverage

That's a fairly aggressive removal of what looked to me like dead code, as well as a lot more unit test so we could boost the coverage. It would be a real help to know if I've inadvertently cut anything that should have been left alone. The pull request is #116 

I'll also look into the `distutils` warning and see if we can fix that.

jamiebull1 commented:
@lafrech Just to be clear, are you running these tests on OS X? It looks like an Apple problem for example [here](https://github.com/pypa/pip/issues/3165)), and the simplest "solution" is to run:

    pip install git+https://github.com/santoshphilip/eppy@i87_six --ignore-installed six

I'm not on a Mac though, so can't test that out. Our CI tests on OS X install a fresh version of Python so don't have the same problem with Apple's default Python installation.

lafrech commented:
We just tried both

    pip install git+https://github.com/jamiebull1/eppy@improve_coverage
    pip install git+https://github.com/santoshphilip/eppy@jamiebull1-improve_coverage

and didn't see any regression. Chances are our usage of `eppy` covers less than coverage tests anyway.

We run the tests on Windows. The first time, we had that deprecation warning where the library to uninstall was `simpy`. The second time, we didn't. In the tests earlier today, we had two or more of these warnings and I don't remember which lib was to be uninstalled. I didn't investigate this at all.

jamiebull1 commented:
Ok, thanks again. If not on a Mac, the `distutils` issue is related to what has previously been installed on your system. Installing in a fresh `venv` or `conda env` ought to avoid any warnings as it's not trying to remove any existing versions of dependencies first.

jamiebull1 commented:
Ok, that's fixed. The bug was in the test, not the code. I've reformatted the tests including setup and teardown, and ensured that we're testing encoding and line endings for all options.

The main issue is when you open a file in Python 3, unless you specify to read it as binary/bytes, it loses (I'm sure it would say "fixes") the line endings.

I've also added in a few bits and pieces that I'd been meaning to get to including replacing the `from __future__ import` statements to help with cross-version coding, and adding some some paths in `pytest_helpers` so we can import them consistently (IDD_FILES, IDF_FILES, etc.).

Finally, I've added a environment variable on the CI server which tells `pytest` to run the integration tests. To mark a test as an integration test, add the decorator:

```
@pytest.mark.skipif(
    not do_integration_tests(), reason="$EPPY_INTEGRATION env var not set")  # can change the reason if required
```

And to run integration tests on your local machine add the environment variable `EPPY_INTEGRATION=TRUE`.

jamiebull1 commented:
I'll take a look at the merge conflicts now.

santoshphilip commented:
the changes for the integration tests are very nicely done.
I'll take the next steps

santoshphilip commented:
`i87_six` has been merged into `develop` and then into `master`.
This has been released on pypi as `release0.5.40`

lafrech commented:
Thanks and congratulation, guys!

santoshphilip commented:
that was mostly done by @jamiebull1 working on `six`. 
great job !

jamiebull1 commented:
Thanks @santoshphilip! Doubt it would have happened without your networking skills turning up `six` in the first place :-)
